examples/psp22-extension/runtime/psp22-extension-example.rs

```diff
@@ -0,0 +1,398 @@
+use codec::{
+    Decode,
+    Encode,
+    MaxEncodedLen,
+};
+use frame_support::{
+    dispatch::RawOrigin,
+    log::{
+        error,
+        trace,
+    },
+    pallet_prelude::*,
+    traits::fungibles::{
+        approvals::{
+            Inspect as AllowanceInspect,
+            Mutate as AllowanceMutate,
+        },
+        Inspect,
+        InspectMetadata,
+        Transfer,
+    },
+};
+use pallet_assets::{
+    self,
+    WeightInfo,
+};
+use pallet_contracts::chain_extension::{
+    ChainExtension,
+    Environment,
+    Ext,
+    InitState,
+    RetVal,
+    SysConfig,
+    UncheckedFrom,
+};
+use sp_runtime::{
+    traits::{
+        Saturating,
+        StaticLookup,
+        Zero,
+    },
+    DispatchError,
+};
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22BalanceOfInput<AssetId, AccountId> {
+    pub asset_id: AssetId,
+    pub owner: AccountId,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22AllowanceInput<AssetId, AccountId> {
+    pub asset_id: AssetId,
+    pub owner: AccountId,
+    pub spender: AccountId,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22TransferInput<AssetId, AccountId, Balance> {
+    pub asset_id: AssetId,
+    pub to: AccountId,
+    pub value: Balance,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22TransferFromInput<AssetId, AccountId, Balance> {
+    pub asset_id: AssetId,
+    pub from: AccountId,
+    pub to: AccountId,
+    pub value: Balance,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22ApproveInput<AssetId, AccountId, Balance> {
+    pub asset_id: AssetId,
+    pub spender: AccountId,
+    pub value: Balance,
+}
+
+pub struct Psp22Extension;
+
+fn map_err(err_msg: &'static str) -> impl FnOnce(DispatchError) -> DispatchError {
```

Comment:
I'd personally inline this, but if you do want to keep it I'd change the name so we don't end up with slightly odd `map_err(map_err("foo"))` statements

Reply:
Would `convert_err` sound ok to you?

Reply:
Yeah that's fine